docs/apps/hyperqueue.md

```diff
@@ -4,234 +4,280 @@ tags:
+We recommend reading the official [HyperQueue documentation](https://it4innovations.github.io/hyperqueue/stable/).
 
-Specify where on the file system the HyperQueue server should be placed. All `hq` commands
-respect this variable so make sure it's set before you call any `hq` commands. The server
-location can also be placed using the command line flag `--server-dir /server/location/on/lustre`.
+This example consists of a batch script and an executable task script.
+The batch script starts the HyperQueue server and workers and submits tasks to the workers.
+The task script is an executable script that we submit to the workers.
+You can copy the following example, run it as given, and modify it to suit your needs.
+The directory structure looks as follows:
 
-```bash
-export HQ_SERVER_DIR=/server/location/on/lustre
+```text
+.             # Current working directory
+├── batch.sh  # Batch script for HyperQueue server and workers
+└── task      # Executable task script for HyperQueue
 ```
 
-If the server directory is not specified it will default to the user home directory. In this case
-one has to be careful not to mix up separate computations. For simple cases that fit inside one
-Slurm job, we recommend starting one server per job in some job-specific directory.
+**Task**
 
-Start the server:
+We assume that HyperQueue tasks are independent and run on a single node.
+Here is an example of a simple, executable `task` script written in Bash.
 
 ```bash
-hq server start &
-until hq job list &>/dev/null ; do sleep 1 ; done
+#!/bin/bash
+sleep 1
 ```
 
-Here, the server is placed in the background (`&`) so that we can continue.
-
-### Starting the workers
-
-Start the workers (again in the background so our script can continue):
+The overhead per task is around 0.1 milliseconds.
+Therefore, we can efficiently execute even very small tasks.
+
+**Batch job**
+
+In a Slurm batch job, each Slurm task corresponds to one HyperQueue worker.
+We can increase the number of workers by increasing the number of Slurm tasks.
+We reserve a fraction of the CPUs and memory on a node per worker in a partial node allocation and all the CPUs and memory on a node per worker in a full node allocation.
+
+=== "Puhti partial single node"
+    ```bash
+    #!/bin/bash
+    #SBATCH --account=<project>
+    #SBATCH --partition=small    # single node partition
+    #SBATCH --ntasks=1           # one HyperQueue worker
+    #SBATCH --cpus-per-task=10   # one or more cpus per worker
+    #SBATCH --mem-per-cpu=1000   # desired amount of memory
+    #SBATCH --time=00:15:00
+    ```
+
+=== "Puhti partial multinode"
+    ```bash
+    #!/bin/bash
+    #SBATCH --account=<project>
+    #SBATCH --partition=large    # multi node partition
+    #SBATCH --ntasks=2           # two or more HyperQueue workers
+    #SBATCH --cpus-per-task=10   # one or more cpus per worker
+    #SBATCH --mem-per-cpu=1000   # desired amount of memory
+    #SBATCH --time=00:15:00
+    ```
```

Comment:
Perhaps
```bash
#SBATCH --nodes=2
#SBATCH --ntasks-per-node=1
```
would be more clear?

Reply:
More generally, why reserve 2 nodes with 10 threads instead of 1 node with 20 threads?

Reply:
I was thinking that it is easier for Slurm to allocate a smaller fraction of CPUs on a node to more tasks (processes) so that slurm can decide where to place them.
Since HyperQueue tasks don't communicate with one another, it shouldn't matter if they run of workes on different nodes.